### PR TITLE
Detect sap workload

### PR DESCRIPTION
**Pull request by ngetahun**

### CODE REVIEW

**Comment by ngetahun** [comment]
[[CAT] Detect SAP workloads](https://trello.com/c/tJPmHP8l/3392-cat-detect-sap-workloads)

File: internal/collectors/sap.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		systemPath := getAbsPath(sapInstallationDir, systemId)
+		workloads, _ := getMatchedSubdirectories(systemPath, workloadsRegex)
+
+		detector = append(detector, SAPDetected{
+			systemId:      systemIds[0],
+			instanceTypes: workloads,
+		})
+	}
+
+	return Result{"sap": detector}, nil
```
[the remark this answers is not in the document]
Comment (ngetahun) [reply internal/collectors/sap.go:69]: We assume the presence of `/usr/sap` signals a full installation of sap.

File: internal/collectors/sap.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	systemIds, err := getMatchedSubdirectories(sapInstallationDir, sapSystemId)
+
+	if err != nil {
+		return NoResult, err
+	}
+
+	var detector []SAPDetected
+	for _, systemId := range systemIds {
+		systemPath := getAbsPath(sapInstallationDir, systemId)
```
[the remark this answers is not in the document]
Comment (ngetahun) [reply internal/collectors/sap.go:60]: good idea

File: internal/collectors/sap.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+type SAP struct{}
+
+// sap follows directory structure of /usr/sap
+// systemId and workloadId will be under the main directory
+// eg /usr/sap/AB3/ERS1
+var sapSystemId = regexp.MustCompile("([A-Z][A-Z0-9]{2})")
+var workloadsRegex = regexp.MustCompile("([A-Z]+)[0-9]{2}")
+var localOsReaddir = os.ReadDir
+
+var sapInstallationDir = "/usr/sap"
```
[the remark this answers is not in the document]
Comment (ngetahun) [reply internal/collectors/sap.go]: done

File: internal/collectors/sap.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixsch) [reply internal/collectors/sap.go:64]: fair enough

**Review by felixsch — APPROVED** [review]
lgtm :+1: 

image


**How I reviewed this merge request:**

- Checked if I would suggest improvements for the provided changes
- Ran connect with `/usr/sap/DEV/{ASC01, DVEBMGS02}` and checked that the output matches
- Ran connect without `/usr/sap` and checked that everything works
- Ran connect and registered against RMT then created the directories and triggered keepalive and checked on RMT that the system is available and includes SAP information.

cheers,

File: internal/collectors/sap.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		return NoResult, err
+	}
+
+	var detector []map[string]interface{}
+	for _, systemId := range systemIds {
+		systemPath := path.Join(sapInstallationDir, systemId)
+		workloads, _ := getMatchedSubdirectories(systemPath, workloadsRegex)
+		detector = append(detector, map[string]interface{}{
+			"systemId":      systemId,
+			"instanceTypes": workloads,
```
Comment (felixsch) [inline internal/collectors/sap.go:54]: We have everything in snake case, I would not start with camel case here :)
```suggestion
			"system_id":      systemId,
			"instance_types": workloads,
```

**Pull request merged**